### PR TITLE
Static complexity, lexical search and more

### PR DESCRIPTION
**Pull request by jmpounders**
There are four new tools in this PR:

1. `CyclomaticComplexityTool` computes the (cyclomatic) complexity of either (1) a file, or (2) the whole codebase, returning only the most complex files for summarization.
2. `FindUsageTool` finds all of the calls to/uses of an object (class, function, variable).  This is currently done by examining imports, so use of an object within the file where it is defined is ignored.  Specifying what object you want to search for must also be precise, e.g., to find uses of `make_react_agent` you have to specify the full "path" to the object, `neosophia.agents.react_chat.make_react_agent`.  This is potentially difficult, but there is not a good "fuzzy" solution at the moment that will cover both external and internal modules (e.g., "Find where 'torch' is used." and "Find where the `make_agent` function from `some_file.py` is used.")
3. `GetFileContentsTool` simply dumps the contents of a file to context.
4. `InstallationAndUsageTool` will search for documentation for the answer to a user question.

Subset of questions I've tested:
1. What are the top 5 most complex functions in the codebase?
2. What is the complexity of openaiapi.py?
3. Describe where neosophia.agents.react_chat.make_react_agent is used.
4. Where is `torch` used?
5. Show me the contents of the README file.
6. How do I run the bank_agent example?

### CODE REVIEW

File: src/groundcrew/tools.py [hunk 1]
```diff
@@ -350,3 +370,298 @@ def __call__(self, user_prompt: str, include_code: bool) -> str:
...
+            files = {
+                id_: metadata['text']
+                for id_, metadata in zip(all_items['ids'], all_items['metadatas'])
+                if id_[-3:] == '.py'
+            }
+
+        else:
+            files = get_python_files(self.collection)
+
+        sort_on = sort_on.lower()
```
Comment (cameronfabbri) [inline src/groundcrew/tools.py:428]: Suggest changing this to `sort_on = str(sort_on).lower()`

The LLM passed `None` (not sure how?), which resulted in the error:

```
How complex is openaiapi.py?
⠋ Running CyclomaticComplexityToolPlease standby while I run the tool CyclomaticComplexityTool...
("Your task is to analyze the cyclomatic complexity summary of 'openaiapi.py' and provide the complexity information of the file.", {'filepath_inexact': 'openaiapi.py', 'sort_on': None})

sort_on = sort_on.lower()
              ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'lower'
```

File: src/groundcrew/tools.py [hunk 2]
```diff
@@ -350,3 +370,298 @@ def __call__(self, user_prompt: str, include_code: bool) -> str:
         prompt += f'{user_prompt}\n\n'
 
         return self.llm(prompt)
+
+
+class CyclomaticComplexityTool:
```
Comment (cameronfabbri) [inline src/groundcrew/tools.py:375]: `get_complexity` and `summarize` don't use anything from `self`, so they technically could be moved to a utils file or something. I do like keeping them in the class for organization though, so we could add `@staticmethod` to them?

File: src/groundcrew/tools.py [hunk 3]
```diff
@@ -350,3 +370,298 @@ def __call__(self, user_prompt: str, include_code: bool) -> str:
...
+
+    def complexity_analysis(self, files: dict[str, str], sort_on: str) -> tuple[list[str], dict]:
+        """Analyze the complexity of the codebase."""
+
+        file_complexity = {}
+        current_max = 0
+        for file, source_code in files.items():
+            complexity = self.get_complexity(source_code)
+            if complexity['max'] >= self.min_max_complexity or complexity['max'] > current_max:
+                file_complexity[file] = complexity
```
Comment (cameronfabbri) [inline src/groundcrew/tools.py:466]: Do you want this outside the `if`?

File: src/groundcrew/tools.py [hunk 4]
```diff
@@ -350,3 +370,298 @@ def __call__(self, user_prompt: str, include_code: bool) -> str:
...
+                    summary_str += f'  {name}: {desc["complexity"]}\n'
+                elif desc['object'] == 'class':
+                    summary_str += f'  {name}: {desc["complexity"]}\n'
+                    for method, method_desc in desc['methods'].items():
+                        summary_str += f'    {name}.{method}: {method_desc["complexity"]}\n'
+
+        return summary_str
+
+
+class FindUsageTool:
```
Comment (cameronfabbri) [inline src/groundcrew/tools.py:510]: Same comment about `@staticmethod` for `get_usage` and `summarize_usage`.

File: src/groundcrew/code_utils.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+    If the entity or module is not found, None is returned."""
+    test = importable_object.split('.')
+    calls = []
+    for imp in imports:
+        imp_name = imp.name.split('.')
+        n = min(len(imp_name), len(test))
+        if test[:n] == imp_name[:n]:
+            call = imp.asname
+            if n < len(test):
+                call += '.' + '.'.join(test[n:])
```
[the remark this answers is not in the document]
Comment (jmpounders) [reply src/groundcrew/code_utils.py:76]: Nice

File: src/groundcrew/tools.py [hunk 6]
```diff
@@ -350,3 +370,298 @@ def __call__(self, user_prompt: str, include_code: bool) -> str:
...
+        files = get_python_files(self.collection)
+
+        usages = {}
+        for file, source_code in files.items():
+            file_imports = cu.get_imports_from_code(source_code)
+            if cu.imports_entity(file_imports, importable_object):
+                called_as = cu.import_called_as(file_imports, importable_object)
+                source_without_imports = '\n'.join(
+                    line for line in source_code.split('\n')
+                    if 'import' not in line
```
[the remark this answers is not in the document]
Comment (jmpounders) [reply src/groundcrew/tools.py:555]: I see what you mean.  I'm changing to regex search to match import forms.  Should catch most of the failure cases.

File: src/groundcrew/tools.py [hunk 7]
```diff
@@ -350,3 +370,298 @@ def __call__(self, user_prompt: str, include_code: bool) -> str:
...
+        Args:
+            user_prompt (str): The prompt to process.
+        """
+        query = (
+            "What files contain documentation regarding the installation and "
+            "use of the codebase? Include README, configuration and environment "
+            "setup files."
+        )
+        doc_files = query_codebase(query, self.collection, n_results=15, where={'type': 'file'})
+        doc_files_uids = [chunk.uid for chunk in doc_files if '..' not in chunk.filepath]
```
[the remark this answers is not in the document]
Comment (jmpounders) [reply src/groundcrew/tools.py:651]: Makes sense... after a quick test they both perform the same, at least for neo-sophia.  The fragment query feels leaner, so going with that.

File: src/groundcrew/tools.py (hunk 2 above)
Comment (jmpounders) [reply src/groundcrew/tools.py:375]: Open to discuss this, but going off "vibes", these methods don't feel like static methods to me. E.g., they are not meant for general use outside the tool instance.  `code_utils.cyclomatic_complexity` would be the general purpose function, which is called and manipulated w/in the class.

True that `self` is unused, but the fact that `self` is a required arg to all python class methods is one of my gripes about python in general...  again, open to discuss, but my vote would be to not have them as static methods.

File: src/groundcrew/tools.py (hunk 3 above)
Comment (jmpounders) [reply src/groundcrew/tools.py:466]: Inside the `if` to build up a set of maximally complex objects.  Currently, if an object does not pass the threshold it is not included unless there are no other more complex objects.  I should add to the docstring to clarify this.

File: src/groundcrew/tools.py (hunk 1 above)
Comment (bdzimmer) [reply src/groundcrew/tools.py]: @cameronfabbri I believe `None`s are getting in from `Agent.run_tool` which fills in missing parameters with `None`

File: src/groundcrew/tools.py (hunk 2 above)
Comment (bdzimmer) [reply src/groundcrew/tools.py:376]: @cameronfabbri @jmpounders One solution might be making a private (start with double underscores) static method? The double underscores would indicate not for general use, but static illustrates the lack of dependencies.

File: src/groundcrew/tools.py (hunk 1 above)
Comment (cameronfabbri) [reply src/groundcrew/tools.py]: Oh yeah, I wrote that piece of code 😆

File: src/groundcrew/tools.py [hunk 8]
```diff
@@ -350,3 +370,298 @@ def __call__(self, user_prompt: str, include_code: bool) -> str:
...
+        matched to find an exact file path for the project.
+        """
+
+        # ensure that filepath is a real path of a file in the collection
+        filepath = fuzzy_match_file_path(self.collection, filepath_inexact, 50)
+
+        if filepath is None:
+            return f'Could not find a source file matching `{filepath_inexact}`'
+
+        items = self.collection.get(
```
[the remark this answers is not in the document]
Comment (bdzimmer) [reply src/groundcrew/tools.py:617]: Bump

**Review by bdzimmer — APPROVED** [review]
One comment I had about filtering db results based on type looks like it hadn't been addressed but otherwise looks good.

I also noticed that `radon` is missing from `requirements.txt`

**Review by cameronfabbri — APPROVED** [review]

**Pull request merged**